zcash_client_sqlite/src/wallet.rs

```diff
@@ -2553,19 +2751,75 @@ pub(crate) fn put_legacy_transparent_utxo<P: consensus::Parameters>(
         ":address": &output.recipient_address().encode(params),
         ":script": &output.txout().script_pubkey.0,
         ":value_zat": &i64::from(Amount::from(output.txout().value)),
-        ":height": &u32::from(output.height()),
+        ":height": &output.height().map_or(-1, i64::from),
     ];
 
     stmt_upsert_legacy_transparent_utxo.query_row(sql_args, |row| row.get::<_, i64>(0).map(UtxoId))
 }
 
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses, although that case should not occur.
+#[cfg(feature = "transparent-inputs")]
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let mut stmt_mark_ephemeral_address_as_used = conn.prepare_cached(
+        "UPDATE ephemeral_addresses SET used_in_tx = :used_in_tx WHERE address = :address",
+    )?;
+
+    let sql_args = named_params![
+        ":used_in_tx": &tx_ref,
+        ":address": encode_transparent_address_p(params, ephemeral_address),
+    ];
+
+    stmt_mark_ephemeral_address_as_used.execute(sql_args)?;
+    Ok(())
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in the given mined transaction (which may or may not be a transaction we sent).
+/// This has no effect if `address` is not one of our ephemeral addresses.
+#[cfg(feature = "transparent-inputs")]
+pub(crate) fn mark_ephemeral_address_as_mined<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let mut stmt_mark_ephemeral_address_as_mined = conn.prepare_cached(
+        "UPDATE ephemeral_addresses SET mined_in_tx = :mined_in_tx WHERE address = :address",
+    )?;
+
+    let sql_args = named_params![
+        ":mined_in_tx": &tx_ref,
+        ":address": encode_transparent_address_p(params, address),
+    ];
+
+    stmt_mark_ephemeral_address_as_mined.execute(sql_args)?;
+    Ok(())
+}
+
 // A utility function for creation of parameters for use in `insert_sent_output`
 // and `put_sent_output`
-fn recipient_params(
-    to: &Recipient<AccountId, Note>,
+fn recipient_params<P: consensus::Parameters>(
+    params: &P,
+    to: &Recipient<AccountId, Note, OutPoint>,
 ) -> (Option<String>, Option<AccountId>, PoolType) {
     match to {
         Recipient::External(addr, pool) => (Some(addr.encode()), None, *pool),
+        Recipient::EphemeralTransparent {
+            receiving_account,
+            ephemeral_address,
+            ..
+        } => (
+            Some(encode_transparent_address_p(params, ephemeral_address)),
```

Comment:
We maybe should not put the address here because then it will be shown to users in history.

Reply:
@nuttycom alternatively suggested that we _always_ put transparent addresses here, no matter their interpretation, because they are always visible in the protocol.

The question then becomes "what is the purpose of showing non-external transparent addresses"? There are (at least) two:
- Inform the user about details they and others could learn.
  - This is IMO reasonable from a "privacy education" perspective.
- Enable the user to do something with the addresses.
  - This is IMO generally a Bad Idea, especially for ephemeral addresses; we don't want users giving them out like external addresses or sending to/from them manually.

Reply:
This is still unresolved.

Reply:
I don't know whether it actually will be shown in history; we should try to figure that out. I seem to remember that we concluded that it would not be returned in some query that we were looking at when reviewing #1402.

Reply:
I'm happy keeping this as it is here.